sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala

```diff
@@ -70,8 +70,10 @@ object Canonicalize {
 
   /** Rearrange expressions that are commutative or associative. */
   private def expressionReorder(e: Expression): Expression = e match {
-    case a: Add => orderCommutative(a, { case Add(l, r) => Seq(l, r) }).reduce(Add)
-    case m: Multiply => orderCommutative(m, { case Multiply(l, r) => Seq(l, r) }).reduce(Multiply)
+    case a @ Add(_, _, c) =>
```

Comment:
This needs further updates as we can't reorder consecutive `Add`s with different `failOnError` flags. It's OK for now as the `Add`s in a query must have the same flag. Can we add a TODO here?